core/src/main/scala/org/apache/spark/sql/SnappySession.scala

```diff
@@ -1014,16 +1014,19 @@ class SnappySession(@transient private val sc: SparkContext,
               }
             }
           case None =>
-            val r = DataSource(self,
+            val dataSource = DataSource(self,
               className = source,
               userSpecifiedSchema = userSpecifiedSchema,
               partitionColumns = partitionColumns,
-              options = params).write(mode, data)
-            if (None != userSpecifiedSchema) {
-              (r, Some(userSpecifiedSchema.get))
-            } else {
-              (r, Some(r.schema))
+              options = params)
+            val clazz: Class[_] = dataSource.providingClass
+            val r = clazz.newInstance() match {
+              case schemaRelationProvider: SchemaRelationProvider =>
+                schemaRelationProvider.createRelation(sqlContext, new CaseInsensitiveMap(params)
+                  , userSpecifiedSchema.getOrElse(data.schema))
```

Comment:
What happens for other cases that do not match SchemaRelationProvider (e.g. column tables)? It will fail with MatchError.

I think there are two options here:

a) Apply new schema to the DataFrame if the two don't match:
```scala
data = userSpecifiedSchema match {
  case Some(s) if (s != data.schema) => session.internalCreateDataFrame(data.queryExecution.toRdd, s)
  case _ => data
}
```
Then go ahead with normal DataSource.write call which will create with updated schema.

b) Alternatively use the SchemaRelationProvider to just do the create, and then let the write go ahead as above. In this case the creation and write will not be atomic in the sense that if write fails then table will not be automatically destroyed (e.g. see last lines in ColumnFormatRelation). This can be handled by checking for "r" against DestroyRelation and invoking its destroy if write fails.

Also discuss with Hemant since he is changing the order of hive catalog entry vs table creation+write.